### PR TITLE
Windows eventlog pattern parser

### PR DESCRIPTION
**Pull request by susu**
Here are my windows eventlog patterndb related modifications.

- I'm not sure about the place of the PatternDB XML file.
- I've only added the new SCL directory to `Makefile.am`, but not to `CMakeLists.txt`. Is it OK?
- I'm stuck with the deployment of the PatternDB XML file. How should I deploy it to `/etc/patterndb.d/...`?
- I'm not sure how I should integrate executing PatternDB test (`pdbtool test --validate ...`)

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build? (admin please type: ok to test)

**Comment by bazsi** [comment]
The patternsb file should be in scl, even when installed. You can then use db-parser() file argument to refer to it.

**Comment by bazsi** [comment]
@kira-syslogng: ok to test

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

File: scl/windowspatternparser/windows.conf [hunk 1]
```diff
@@ -0,0 +1,23 @@
+
+block parser windows-eventlog-parser() {
+    channel {
+        filter {
+            match("(4624|4634)" value(".SDATA.win@18372\.4.EVENT_ID"));
+        };
+        parser {
```
Comment (bazsi) [inline scl/windowspatternparser/windows.conf]: do we need the filter here? db-parser() should be good enough for classification whether the event was identified, so later we can filter on tags, which should be faster.

For this use-case, it might actually make sense to drop unmatched messages in db-parser(), hmmm. sending a pull request in a minute.

File: scl/windowspatternparser/windows.conf [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    template("$(format-json --scope nv-pairs --exclude MESSAGE \
+                --exclude SOURCE \
+                win.EventRecordID=${.SDATA.win@18372\.4.EVENT_REC_NUM} \
+                win.TimeCreated=${ISODATE} \
+                win.Channel=${.SDATA.win@18372\.4.EVENT_NAME} \
+                win.EventID=${.SDATA.win@18372\.4.EVENT_ID} \
+                PATTERN_CLASS=${.classifier.class} \
+               )\n");
+};
+
```
Comment (bazsi) [inline scl/windowspatternparser/windows.conf]: Where is this template used? Why is it part of this SCL?

File: pattern/windows/windows.xml [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                             <test_value name='win.ProcessId'>0xabc</test_value>
+                             <test_value name='win.LogonType'>10</test_value>
+                             <test_value name='win.TargetDomainName'>BSP-WIN2K8R2-EN</test_value>
+                             <test_value name='win.LogonGuid'>{00000000-0000-0000-0000-000000000000}</test_value>
+                             <test_value name='win.AuthenticationPackageName'>Negotiate</test_value>
+                             <test_value name='win.LogonProcessName'>User32 </test_value>
+                             <test_value name='win.LmPackageName'>-</test_value>
+                             <test_value name='win.TransmittedServices'>-</test_value>
+                             <test_value name='win.KeyLength'>0</test_value>
+                         </test_values>
```
Comment (bazsi) [inline pattern/windows/windows.xml]: name value pairs should start with "." to indicate a reserved key-value namespace.

File: scl/windowspatternparser/windows.conf (hunk 1 above)
Comment (bazsi) [inline scl/windowspatternparser/windows.conf]: And here is the pull request: https://github.com/balabit/syslog-ng/pull/1574

File: scl/Makefile.am [hunk 4]
```diff
@@ -15,7 +15,8 @@ SCL_SUBDIRS	= \
 	cisco		\
 	loggly		\
 	logmatic	\
-	snmptrap
+	snmptrap	\
+	windowspatternparser
 SCL_CONFIGS	= scl.conf syslog-ng.conf
```
Comment (presidento) [inline scl/Makefile.am]: Please call it `windowseventlogdbparser`. We want to keep the short name for the "official" objects, which will be coming soon. (For more information please turn to Varnyu.)

Comment (bazsi) [inline scl/Makefile.am]: Why is it not an official parser? Those can come in the same directory. Anyway, Ill try to talk to him IRL.

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by bazsi** [comment]
Travis failed because the files are not properly included in make dist, and thus distcheck fails.

**Comment by presidento** [comment]
There is a silence here, moved to purgatory.

**Comment by gorosz** [comment]
Can you please review this once again.
I hope, I do all necessary changes from the reviews

File: scl/windowspatternparser/windows.conf (hunk 2 above)
Comment (bazsi) [reply scl/windowspatternparser/windows.conf]: You still have the template here, what's the use-case, why is that needed here?

File: scl/windowseventlog/plugin.conf [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+template t_wineventlog_json {
+    template("$(format-json --scope nv-pairs --exclude MESSAGE \
+                --exclude SOURCE \
+                win.EventRecordID=${.SDATA.win@18372\.4.EVENT_REC_NUM} \
+                win.TimeCreated=${ISODATE} \
+                win.Channel=${.SDATA.win@18372\.4.EVENT_NAME} \
+                win.EventID=${.SDATA.win@18372\.4.EVENT_ID} \
+                PATTERN_CLASS=${.classifier.class} \
+               )\n");
```
Comment (presidento) [inline scl/windowseventlog/plugin.conf]: > You still have the template here, what's the use-case, why is that needed here?

by @bazsi, from https://github.com/balabit/syslog-ng/pull/1572#discussion_r138559874

Comment (bazsi) [reply scl/windowseventlog/plugin.conf]: I've talked to Gergo and agreed that the template would be removed, with that this is ready to be merged.

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

File: scl/windowseventlog/plugin.conf (hunk 5 above)
Comment (gorosz) [reply scl/windowseventlog/plugin.conf]: As requested removed the template from the config

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions
